### PR TITLE
Improve notes view UI and fix introduced bug

### PR DESCRIPTION
**Pull request by Atlas139mkm**
Please give lots of advice as I want to ensure what I have done is up to standard.

### Description of the Change

1:
Altered the notes view to remove the need for the pop up while retaining and improving existing usability and function.
- no more pop up
- added "add selected" and "remove selected" button to the note
- altered functionality of "edit" button to allow editing of the note directly instead of via a pop up
- note saves if its in edit or view state, even if the user closes notes view or changes the nodes/transactions it's linked to.
- altered linked to text to state "node" and "transaction" if there is only 1 element, "nodes" and "transactions" if there are multiple (requested in #1178)

2:
- fixed an issue from #1316 that was merged to master today, where creating a new graph would carry notes over from the previously created graph to the new one.

Main changes in the code are new buttons, removed pop up, added a variable to NotesViewEntry that states if the note is in edit or view mode and fixed the #1316 issue (this can be seen in master at the moment).

### Preview of New Notes View
Notes video demo

### Alternate Designs

- Potentially adding right click to edit the text areas
- Change the colours of the buttons to be a traffic light style (blue edit, green add, orange remove, red delete)
- reducing buttons on the note by combing some (I couldn't think of a way to do this without pop ups)

### Why Should This Be In Core?

- It fixes the recent bug merged
- It improves usability of the notes view
- It maintains all existing functionality

### Benefits

- Increased usability
- removal of bug

### Possible Drawbacks

- The size of the editable text box on the note can sadly not be increased in height
   - There was a solution to this using a InlineCssTextArea, however this slowed creation of the first note by at least 3 seconds (unsure why so I went with speed performance over a dynamically resizing text area)

### Verification Process

- I pressed every combination of buttons on the note I could think of.
- I tested various combinations of notes with no or null text areas
- I imported graphs with existing notes
- I removed and added the selected nodes and transactions while the node was in edit mode
- I made new graphs
- I looked at the other notes features such as the filtering and ensured the notes still appeared 
- I re-tested the same process that caused the original bug for #1316 

### Applicable Issues

#1290
#1178
#1316

### CODE REVIEW

**Comment by antares1470** [comment]
@Atlas139mkm just a request but are you possibly able to add the bug fix as a separate PR? I think it would be good to get that in before v2.4 is released

**Review by sol695510 — APPROVED** [review]

**Review by antares1470 — CHANGES_REQUESTED** [review]
I think it would be nice to find a way to make the notes themselves smaller. While adding the buttons to user notes removes pop ups, it makes the user notes a lot larger than they need to be and therefore requires a user to do more scrolling through notes which is less than ideal. The sizing issue also affects auto notes which definitely isn't ideal.

NotesViewNoteSize

One option for reducing size could be to merge the 'Add Selection' and 'Remove Selection' buttons into one button. This will require adding submenus so that you can pick whether to add or remove. An alternative might be moving the selection change functionality to the right-click menu (which already contains viewing selection on the graph). Just some thoughts

**Comment by Atlas139mkm** [comment]
Thanks for the feedback @antares1470! I'll get onto the changes in a bit. I'll see what I can do about adding the add and remove to a right click option. That seems like a much better way to do it too keep the size down.

File: CoreNotesView/src/au/gov/asd/tac/constellation/views/notes/NotesViewPane.java [hunk 1]
```diff
@@ -595,53 +597,91 @@ private void createNote(final NotesViewEntry newNote) {
...
+        // Define selection label
+        String selectionLabelText = "";
+        final Label selectionLabel = new Label(selectionLabelText);
+
         // If the note is user created add the selection details.
         if (newNote.isUserCreated()) {
-            String selectionLabelText;
             if (newNote.isGraphAttribute()) {
-                selectionLabelText = "Note selection is linked to the graph.";
+                selectionLabelText = "Note is linked to: the graph.";
```
Comment (arcturus2) [inline CoreNotesView/src/au/gov/asd/tac/constellation/views/notes/NotesViewPane.java:633]: How about "Note linked to: Graph" to reduce the amount of words we have to see here on the note?

File: CoreNotesView/src/au/gov/asd/tac/constellation/views/notes/NotesViewPane.java [hunk 2]
```diff
@@ -595,53 +597,91 @@ private void createNote(final NotesViewEntry newNote) {
...
             } else {
-                selectionLabelText = "Note selection is linked to " + newNote.getNodesSelected().size() + " nodes and "
-                        + newNote.getTransactionsSelected().size() + " transactions.";
+                selectionLabelText = "Note is linked to: ";
```
Comment (arcturus2) [inline CoreNotesView/src/au/gov/asd/tac/constellation/views/notes/NotesViewPane.java:635]: Similarly that would then make this "Note linked to: "

File: CoreNotesView/src/au/gov/asd/tac/constellation/views/notes/NotesViewTopComponent.java [hunk 3]
```diff
@@ -116,7 +116,7 @@ protected void handleComponentOpened() {
     @Override
     protected void handleComponentClosed() {
         super.handleComponentClosed();
-        notesViewPane.closeEdit();
+        //notesViewPane.selectAllFilters();
```
Comment (arcturus2) [inline CoreNotesView/src/au/gov/asd/tac/constellation/views/notes/NotesViewTopComponent.java:119]: Given the PR is going into `master` please remove dead commented code if you don't need it.

**Review by arcturus2 — CHANGES_REQUESTED** [review]
@Atlas139mkm this is really neat and thank you very much for thinking how this could be improved and also implementing a neat solution.

I have left some comments for you but besides them, can you please look into adding an "Are you sure?" type message when a user clicks to delete the note? I think in terms of UX this is important. Noted also that I am happy for this to be put into a ticket to be worked on later as this is technically not part of anything introduced in your PR.

Let me know your thoughts.

File: CoreNotesView/src/au/gov/asd/tac/constellation/views/notes/NotesViewPane.java (hunk 1 above)
Comment (Atlas139mkm) [reply CoreNotesView/src/au/gov/asd/tac/constellation/views/notes/NotesViewPane.java:633]: Yep I agree, ill change it.

**Comment by Atlas139mkm** [comment]
Awesome thanks @arcturus2, ill make your requested changes can look into a nice way of displaying an are you sure message.

**Comment by Atlas139mkm** [comment]
Actually thoughts on this?

Delete button would convert to disabled and change to "are you sure?" (I would probably disable the edit button too until they pick yes or no)

no and yes buttons would appear and yes would delete it, no would covert the buttons back to enabled and remove the no and yes buttons.

image

**Comment by arcturus2** [comment]
@Atlas139mkm that's pretty cool, I wouldn't have thought of that as a way to present it. I am open to try this idea out if you like.

For me, I was thinking of going down the traditional pop up dialog using a `new Alert(Alert.AlertType.CONFIRMATION)` to get the OK/Cancel response given that's what I am used to seeing.

**Comment by Atlas139mkm** [comment]
I'm happy to do either way. I figured with this way it avoids more pop ups, the confirmation message happens directly under the users mouse and it keeps everything happening in the one location.

**Comment by arcturus2** [comment]
@Atlas139mkm maybe we should get more opinions on this one.

In terms of the rest of the app there are uses of the `Alert` dialog and checking for a confirmation. I think it is ok for a popup to appear when we really want to get the users attention and it is fine for this scenario to do so.

You got me thinking further and if I think about design principles, I am not sure if showing the "are you sure" in a button is the best way to do it because it still looks like a button. I would instead use a `Label`. But this means more use of objects and setting things visible/invisible...

I totally understand your thinking about using less popups and reusing components which are great ideas.

What do you think @antares1470 @cygnus-x-1  @aldebaran30701 @Delphinus8821 ?

**Comment by Atlas139mkm** [comment]
Yeah I was thinking that, I could also just temporarily remove the edit button, delete note button and replace it with a label saying are you sure? and the no and yes buttons. Takes a little bit more reconfiguring but it would probably look better

**Comment by aldebaran30701** [comment]
I'm a fan - However I think you are right about the button for displaying text. I feel like so the user doesn't just miss the fact that the text changed that it should have a larger text. I'd say just needs to be a bit more noticeable. Many popups will get annoying especially if you have a few notes to clear out.
image

**Comment by sol695510** [comment]
Notes Button layout suggestion
@arcturus2 I remember you pointing me here when I initially started creating the Notes View, I think this method of hiding/unhiding the buttons on hover should be considered for the Notes View, but I can understand it being moved to a new ticket.

Thoughts? @antares1470 @aldebaran30701 @Delphinus8821 @Atlas139mkm

**Comment by antares1470** [comment]
@sol695510 I reckon it could be a cool addition, I think it could be worth a separate ticket for it though.

@Atlas139mkm @arcturus2 For me I think you want to force people to address the delete and that's why the pop up isn't a bad idea. It forces their eye to move to the dialog and they have to acknowledge it before they can do anything else. 

What I think we want to avoid is a user accidentally clicking the delete button and continuing on like nothing happened because they haven't noticed the buttons change. That scenario would result in the possibility of them accidentally pressing yes later on which I think leaves you not much better off than what we have currently (i.e. no confirmation check)

**Comment by Atlas139mkm** [comment]
So you would prefer a confirmation pop up over implementing what @aldebaran30701 suggested?

**Comment by antares1470** [comment]
> So you would prefer a confirmation pop up over implementing what @aldebaran30701 suggested?

Regarding pop ups, I think our objective is to avoid having any unnecessary ones, rather than avoiding them altogether. Sometimes, a pop up can actually be a good solution (particularly if confirming actions with notable consequences).

That being said I think the main things to be addressed are that firstly there is confirmation, and secondly that the confirmation needs to be addressed immediately. A pop up like @arcturus2 suggested is probably the easiest way to implement that but if @aldebaran30701's solution also addresses those two points then I'm not against it

**Comment by Atlas139mkm** [comment]
it would satisfy the fact that there is confirmation, however it would not warranty any level of urgency. The user would not be forced to respond to it before doing anything else. They would not be able to do anything else with that specific note but they could leave the notes view and do anything else they liked. I'm also having sizing issues implementing it atm as well since a note dynamically resizes I can capture its current size so it looks a bit strange when you click delete note. I think a pop up might be the way to go if its the very next action you want the user to do.

**Comment by Atlas139mkm** [comment]
Alright I have just committed some more changes. 

This is the note.
image

Confirmation message on delete. will only delete on user clicking OK.
image

Note in edit mode
image

User right click options to add and remove selected graph elements
image

Thoughts?

**Review by antares1470 — CHANGES_REQUESTED** [review]
Looking good. I think the only thing to add now is a whatsnew entry

File: CoreWhatsNewView/src/au/gov/asd/tac/constellation/views/whatsnew/whatsnew.txt [hunk 4]
```diff
@@ -1,6 +1,14 @@
...
 <p>If you're new to Constellation, read the <a href="" helpId="au.gov.asd.tac.constellation.functionality.gettingstarted">getting started guide</a>.</p>
 
+== 2021-09-08 Notes View UI Improvements
+<p>New features:
+<ul>
+<li><p>Replaced edit pop up with direct note editable text fields</p></li>
+<li><p>Right click note to add and remove selected graph elements</p></li>
+<li><p>Delete note confirmation pop up</p></li>
+</ul>
+
```
Comment (antares1470) [inline CoreWhatsNewView/src/au/gov/asd/tac/constellation/views/whatsnew/whatsnew.txt:11]: The first `<p>` is missing a `</p>`. The list items themselves also don't need to be contained in `<p></p>`

**Review by antares1470 — CHANGES_REQUESTED** [review]

**Review by antares1470 — APPROVED** [review]

**Review by arcturus2 — APPROVED** [review]
Works nicely

**Pull request merged**